examples/ex29.cpp

```diff
@@ -0,0 +1,375 @@
+//                                MFEM Example 29
+//
+// Compile with: make ex29
+//
+// Sample runs:  ex29
+//               ex29 -r 2 -sc
+//               ex29 -mt 3 -o 4 -sc
+//               ex29 -mt 3 -r 2 -o 4 -sc
+//
+// Description:  This example code demonstrates the use of MFEM to define a
+//               finite element discretization of a PDE on a 2 dimensional
+//               surface embedded in a 3 dimensional domain. In this case we
+//               solve the Laplace problem -Div(sigma Grad u) = 1, with
+//               homogeneous Dirichlet boundary conditions, where sigma
+//               is an anisotropic diffusion constant defined as a 3x3 matrix
+//               coefficient.
+//
+//               This example demonstrates the use of finite element
+//               integrators on 2D domains with 3D coefficients.
+//
+//               We recommend viewing example 1 before viewing this example.
+
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+
+using namespace std;
+using namespace mfem;
+
+Mesh * GetMesh(int type);
+
+void trans(const Vector &x, Vector &r);
+
+void sigmaFunc(const Vector &x, DenseMatrix &s);
+
+double uExact(const Vector &x)
+{
+   return (0.25 * (2.0 + x[0]) - x[2]) * (x[2] + 0.25 * (2.0 + x[0]));
+}
+
+void duExact(const Vector &x, Vector &du)
+{
+   du.SetSize(3);
+   du[0] = 0.125 * (2.0 + x[0]) * x[1] * x[1];
+   du[1] = -0.125 * (2.0 + x[0]) * x[0] * x[1];
+   du[2] = -2.0 * x[2];
+}
+
+void fluxExact(const Vector &x, Vector &f)
+{
+   f.SetSize(3);
+
+   DenseMatrix s(3);
+   sigmaFunc(x, s);
+
+   Vector du(3);
+   duExact(x, du);
+
+   s.Mult(du, f);
+   f *= -1.0;
+}
+
+int main(int argc, char *argv[])
+{
+   // 1. Parse command-line options.
+   int order = 3;
+   int mesh_type = 4; // Default to Quadrilateral mesh
+   int ref_levels = 0;
+   bool static_cond = false;
+   bool visualization = true;
+
+   OptionsParser args(argc, argv);
+   args.AddOption(&mesh_type, "-mt", "--mesh-type",
+                  "Mesh type: 3 - Triangular, 4 - Quadrilateral.");
+   args.AddOption(&ref_levels, "-r", "--refine",
+                  "Number of times to refine the mesh uniformly in serial.");
+   args.AddOption(&order, "-o", "--order",
+                  "Finite element order (polynomial degree) or -1 for"
+                  " isoparametric space.");
```

Comment:
The mesh is constructed by the example, so "isoparametric" is not well defined. In examples where we read the mesh from a file "isoparametric" space means to use the curvature/order of the input mesh.

Reply:
Very true. I'll fix this.